pycrostates/preprocessing/spatial_filter.py

```diff
@@ -0,0 +1,189 @@
+from mne.io import BaseRaw
+from mne.io.pick import _picks_by_type
+from mne.parallel import parallel_func
+from mne.utils.check import _check_preload
+from numpy.typing import NDArray
+
+from .._typing import CHData
+from ..utils._checks import _check_n_jobs, _check_type, _check_value
+from ..utils._docs import fill_doc
+from ..utils._logs import logger, verbose
+
+
+@fill_doc
+@verbose
+def apply_spatial_filter(
+    inst: Union[BaseRaw, BaseEpochs, CHData],
+    ch_type: str = "eeg",
+    exclude_bads: bool = True,
+    origin: Union[str, NDArray[float]] = "auto",
+    n_jobs: int = 1,
+    verbose=None,
+):
+    r"""Apply a spatial filter.
+
+    Adapted from \ :footcite:t:`michel2019eeg`.
+    Apply an instantaneous filter which interpolates channels
+    with local neighbors while removing outliers.
+    The current implementation proceeds as follows:
+
+    * An interpolation matrix is computed using
+      ``mne.channels.interpolation._make_interpolation_matrix``.
+    * An ajdacency matrix is computed using
+      `mne.channels.find_ch_adjacency`.
+    * If ``exclude_bads`` is set to ``True``,
+      bad channels are removed from the ajdacency matrix.
+    * For each timepoint and each channel,
+      a reduced adjacency matrix is built by removing neighbors
+      with lowest and highest value.
+    * For each timepoint and each channel,
+      a reduced interpolation matrix is built by extracting neighbor
+      weights based on the reduced adjacency matrix.
+    * The reduced interpolation matrices are normalized.
+    * The channel's timepoints are interpolated
+      using their reduced interpolation matrix.
+
+    Parameters
+    ----------
+    inst : Raw | Epochs | ChData
+        Instance to filter spatially.
+    ch_type : str
+        The channel type on which to apply the spatial filter.
+        Currently only supports ``'eeg'``.
+    exclude_bads : bool
+        If set to ``True``, bad channels will be removed
+        from the adjacency matrix and therefore not used
+        to interpolate neighbors. In addition, bad channels
+        will not be filtered.
+        If set to ``False``, proceed as if all channels were good.
+    origin : array of shape (3,) | str
+        Origin of the sphere in the head coordinate frame and in meters.
+        Can be ``'auto'`` (default), which means a head-digitization-based
+        origin fit.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    inst : Raw | Epochs| ChData
+        The instance modified in place.
+
+    Notes
+    -----
+    This function requires a full copy of the data in memory.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    _check_type(inst, (BaseRaw, BaseEpochs, CHData), item_name="inst")
+    _check_type(ch_type, (str,), item_name="ch_type")
+    _check_value(ch_type, ("eeg",), item_name="ch_type")
+    _check_type(exclude_bads, (bool,), item_name="exclude_bads")
+    origin = _check_origin(origin, inst.info)
```

Comment:
```suggestion
```

`origin` is now tested twice. I think it makes more sense to test it later with the picked info.